src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs

```diff
@@ -13,25 +13,54 @@ public class GenericModelBinder : IModelBinder
     {
         public async Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
         {
-            var binderType = ResolveBinderType(bindingContext.ModelType);
-            if (binderType != null)
+            var bindingInfo = ResolveGenericBindingInfo(bindingContext.ModelType);
+            if (bindingInfo != null)
             {
-                var binder = (IModelBinder)Activator.CreateInstance(binderType);
+                var binder = (IModelBinder)Activator.CreateInstance(bindingInfo.ModelBinderType);
                 var result = await binder.BindModelAsync(bindingContext);
 
-                var modelBindingResult = result != null ?
-                    new ModelBindingResult(result.Model, result.Key, result.IsModelSet, result.ValidationNode) :
-                    new ModelBindingResult(model: null, key: bindingContext.ModelName, isModelSet: false);
-
-                // Were able to resolve a binder type.
+                if (result != null && result.IsModelSet)
+                {
+                    // Success - propegate the values returned by the model binder.
+                    return new ModelBindingResult(result.Model, result.Key, result.IsModelSet, result.ValidationNode);
+                }
+
+                // If this is the fallback case, and we didn't bind as a top-level model, then generate a
+                // default 'empty' model and return it.
+                var isTopLevelObject = bindingContext.ModelMetadata.ContainerType == null;
+                var hasExplicitAlias = bindingContext.BinderModelName != null;
+
+                if (isTopLevelObject && (hasExplicitAlias || bindingContext.ModelName == string.Empty))
+                {
+
+                    var model = result?.Model;
+                    if (model == null && bindingInfo.UnderlyingModelType.IsArray)
+                    {
+                        model = Array.CreateInstance(bindingInfo.UnderlyingModelType.GetElementType(), 0);
+                    }
+                    else if (model == null)
+                    {
+                        model = Activator.CreateInstance(bindingInfo.UnderlyingModelType);
+                    }
+
+                    var validationNode = new ModelValidationNode(
+                        bindingContext.ModelName,
+                        bindingContext.ModelMetadata,
+                        model);
+
+                    return new ModelBindingResult(model, bindingContext.ModelName, true, validationNode);
+                }
+
+                // We always want to return a result for model-types that we handle.
+                //
                 // Always tell the model binding system to skip other model binders i.e. return non-null.
```

Comment:
The existing comment and what you added two lines above look much more closely-related than the separation implies. Suggest merging into one paragraph. Something like "We always want ... This tells the model binding system ..." (and remove the "i.e." bit).